### PR TITLE
feat: Add crowd.dev node and trigger node

### PR DESCRIPTION
**Pull request by perseus-algol**
Adds new node and trigger node crowd.dev

### CODE REVIEW

**Comment by janober** [comment]
Thanks a lot for the submission. I always thought the company is called "crowd.dev", is there a reason why credentials + nodes have the name "crowd" and not rather "crowdDev"?

It also looks like the trigger node is using the old and deprecated request-based helper `requestWithAuthentication` instead of the new `httpRequestWithAuthentication`.

**Comment by perseus-algol** [comment]
@janober, Thanks for the review, the specified issues have been fixed.

File: packages/nodes-base/nodes/CrowdDev/CrowdDevNode.node.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+	"node": "n8n-nodes-base.crowd",
```
Comment (mcnaveen) [inline packages/nodes-base/nodes/CrowdDev/CrowdDevNode.node.json:2]: Like @janober said in https://github.com/n8n-io/n8n/pull/6082#issuecomment-1537916247 Is this supposed to be `crowdDev` right?

**Comment by Joffcom** [comment]
Hey @perseus-algol,

Just noticed this is now ready for review, I have it on my list for Monday 👍🏻

File: packages/nodes-base/credentials/CrowdDevApi.credentials.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	name = 'crowdDevApi';
+
+	displayName = 'crowd.dev API';
+
+	documentationUrl = 'https://docs.crowd.dev/reference/getting-started-with-crowd-dev-api';
+
+	properties: INodeProperties[] = [
+		{
+			displayName: 'Domain',
+			name: 'domain',
```
Comment (janober) [inline packages/nodes-base/credentials/CrowdDevApi.credentials.ts:18]: Should probably be more "url" and not "domain" as it does include "https"

File: packages/nodes-base/nodes/CrowdDev/CrowdDevNode.node.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import type { INodeType, INodeTypeDescription } from 'n8n-workflow';
+import { allProperties } from './descriptions';
+
+export class CrowdDevNode implements INodeType {
```
Comment (janober) [inline packages/nodes-base/nodes/CrowdDev/CrowdDevNode.node.ts:4]: Should not be "crowdDevNode" just "crowdDev". Is obviously the same for the name, defaults, context file, ...

File: packages/nodes-base/nodes/CrowdDev/CrowdDevNode.node.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{
+	"node": "n8n-nodes-base.crowdDev",
```
Comment (janober) [inline packages/nodes-base/nodes/CrowdDev/CrowdDevNode.node.json:2]: That has to actually match with the name of the node. So is wrong right now, but will be correct once the node name got fied.

File: packages/nodes-base/credentials/CrowdDevApi.credentials.ts (hunk 2 above)
Comment (Joffcom) [reply packages/nodes-base/credentials/CrowdDevApi.credentials.ts:18]: Would also add an option for self signed certificates as there will no doubt be a could of users not using Let's Encrypt or similar.

**Comment by perseus-algol** [comment]
I was made specified renames. 
I am not quite sure about an option for self signed certificates. Did you meant to add this one:
```js
{
  displayName: 'Ignore SSL Issues',
  name: 'allowUnauthorizedCerts',
  type: 'boolean',
  default: false,
  description: 'Whether to connect even if SSL certificate validation is not possible',
},
```
?

**Comment by Joffcom** [comment]
@perseus-algol that is the one, then in the node you would set the option based on that value.

**Review by Joffcom — COMMENTED** [review]
Left a few notes for some things to change.

File: packages/nodes-base/credentials/CrowdDevApi.credentials.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+				Authorization: '={{"Bearer " + $credentials.token}}',
+			},
+		},
+	};
+
+	// The block below tells how this credential can be tested
+	test: ICredentialTestRequest = {
+		request: {
+			method: 'POST',
+			baseURL: '={{$credentials?.domain + "/api/tenant/" + $credentials?.tenantId}}',
```
[the remark this answers is not in the document]
Comment (perseus-algol) [reply packages/nodes-base/credentials/CrowdDevApi.credentials.ts]: Just checked API, it works with two slashes also. But if needed, I can add replace: `'={{$credentials?.url.replace(/\\/$/, "") + "/api/tenant/" + $credentials?.tenantId}}'`

File: packages/nodes-base/credentials/CrowdDevApi.credentials.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+	displayName = 'crowd.dev API';
+
+	documentationUrl = 'https://docs.crowd.dev/reference/getting-started-with-crowd-dev-api';
+
+	properties: INodeProperties[] = [
+		{
+			displayName: 'URL',
+			name: 'url',
+			type: 'hidden',
```
[the remark this answers is not in the document]
Comment (perseus-algol) [reply packages/nodes-base/credentials/CrowdDevApi.credentials.ts:19]: There is no self-hosted API.

File: packages/nodes-base/nodes/CrowdDev/descriptions/memberFields.ts [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+	shared.emailsField,
+	{
+		displayName: 'Joined At',
+		name: 'joinedAt',
+		description: 'Date of joining the community',
+		type: 'dateTime',
+		default: '',
+	},
+	{
+		displayName: 'Organizations',
```
[the remark this answers is not in the document]
Comment (perseus-algol) [reply packages/nodes-base/nodes/CrowdDev/descriptions/memberFields.ts:111]: Moved all non required fields into additionalOptions collection, as mentioned above. Left commonFields for required fields at top level.

File: packages/nodes-base/credentials/CrowdDevApi.credentials.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Joffcom) [reply packages/nodes-base/credentials/CrowdDevApi.credentials.ts:19]: @perseus-algol I have checked and it looks like the API is available on self hosted installations.

[the remark this answers is not in the document]
Comment (perseus-algol) [reply packages/nodes-base/credentials/CrowdDevApi.credentials.ts]: Indeed. Sorry I missed it. Changed URL field from hidden to string.

**Comment by Joffcom** [comment]
@perseus-algol looking good, I will give it another proper look over tomorrow morning and see if we can can get this into next weeks release.

**Comment by Joffcom** [comment]
Hey @perseus-algol,

Just finishing up the docs side, This all looks good although one thing that confuses me as a user is Activity > Create or Update with a member.

What is this actually for?
image

I am able to add multiple key / values for a username but I don't know what those fields should actually be, I can also have multiple emails in there.

**Comment by garrrikkotua** [comment]
Hi @Joffcom 

This is needed because in crowd.dev member can have a usernames at several platforms. E.g twitter and github. And here users will submit "twitter: oneusername, github: anotherusername". So we leave completely up to users, which platforms and usernames they want to include. But I believe some description can be added here, showing that it's key-value pairs of platform name (e.g twitter, github, etc) and username on this platform.

Regarding emails - it is an array. In case if a member has several emails associated with him.

Hope that helps!

**Comment by Joffcom** [comment]
@garrrikkotua That would work, Maybe calling it "platform" instead of key for the display, Would you typically see people adding all of the options at one? From what I can see in the API docs only username is required for the member object so maybe the others can be under an additional field within that section so it doesn't look so confusing.

**Comment by Joffcom** [comment]
Hey @garrrikkotua,

By keeping it like it is and showing the optional fields it won't fit our normal UI pattern for nodes, It wouldn't need to be in a new section just moved under an additional options or extra fields button.

So right now to get to that I have press Choose for the Member then we see the 4 different member options, My assumption looking at it would be that I need to complete all of those fields which doesn't appear to be the case. We could have display name, Emails and Joined At as other options that can be added as a user needs them to keep things looking clean.

**Comment by perseus-algol** [comment]
image
These fields are required. Does't that mean that they should stay as they are and shouldn't be moved into "Additional Options"?

**Comment by Joffcom** [comment]
Hey @perseus-algol,

Yeah anything required should be visible.

**Review by Joffcom — DISMISSED** [review]

**Comment by Joffcom** [comment]
Morning,

Looks like we have missed the `0.235.0` window by about 20 minutes so this node will be released in `0.236.0` next week.

**Review by Joffcom — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@0.236.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%400.236.0)

**Pull request merged**